qiskit/transpiler/passes/basis/basis_translator.py

```diff
@@ -77,19 +81,36 @@ def run(self, dag):
         Returns:
             DAGCircuit: translated circuit.
         """
-
-        if self._target_basis is None:
+        if self._target_basis is None and self._target is None:
             return dag
 
+        qarg_indices = {qubit: index for index, qubit in enumerate(dag.qubits)}
         # Names of instructions assumed to supported by any backend.
-        basic_instrs = ["measure", "reset", "barrier", "snapshot", "delay"]
-
-        target_basis = set(self._target_basis).union(basic_instrs)
-
-        source_basis = set()
-        for node in dag.op_nodes():
-            if not dag.has_calibration_for(node):
-                source_basis.add((node.name, node.op.num_qubits))
+        if self._target is None:
+            basic_instrs = ["measure", "reset", "barrier", "snapshot", "delay"]
+            target_basis = set(self._target_basis)
+            source_basis = set()
+            for node in dag.op_nodes():
+                if not dag.has_calibration_for(node):
+                    source_basis.add((node.name, node.op.num_qubits))
+            qarg_with_incomplete = {}
+            incomplete_source_basis = {}
+        else:
+            basic_instrs = ["barrier", "snapshot"]
+            source_basis = set()
+            target_basis = set(self._target) - set(self._incomplete_basis)
+            qarg_with_incomplete = defaultdict(set)
+            for gate in self._incomplete_basis:
+                for qarg in self._target[gate]:
+                    qarg_with_incomplete[qarg].add(gate)
```

Comment:
I may well be wrong, but this _feels_ like it's doing some extra iteration - it feels like there'll be a way to combine it into `_compute_incomplete_basis` and reuse the set differences we're already calculating in there, rather than needing to duplicate it.